File: runtime/python/_pagination.py
# Auto-iterating cursor page. Iterating a SyncPage walks every item across
# pages, fetching lazily; `.items` is just the current page.
from __future__ import annotations

from typing import Callable, Generic, Iterator, List, TypeVar

T = TypeVar("T")


class SyncPage(Generic[T]):
    def __init__(
        self,
        items: List[T],
        next_cursor: str,
        fetch: Callable[[str], "SyncPage[T]"],
    ) -> None:
        self.items = items
        self.next_cursor = next_cursor
        self._fetch = fetch

    def has_next_page(self) -> bool:
        return bool(self.next_cursor)

    def get_next_page(self) -> "SyncPage[T] | None":
        """The next page, or None when this is the last one."""
        if not self.next_cursor:
            return None
        return self._fetch(self.next_cursor)

    def __iter__(self) -> Iterator[T]:
        page: SyncPage[T] | None = self
        while page is not None:
            yield from page.items
            page = page.get_next_page()

File: runtime/ruby/pagination.rb
# frozen_string_literal: true

# Auto-iterating cursor page. `each` walks every item across pages, fetching
# lazily; `items` is just the current page.

module RedwoodModule
  class Page
    include Enumerable

    attr_reader :items, :next_cursor

    def initialize(items, next_cursor, &fetch)
      @items = items
      @next_cursor = next_cursor
      @fetch = fetch
    end

    def next_page?
      !@next_cursor.to_s.empty?
    end

    # The next page, or nil when this is the last one.
    def next_page
      return nil unless next_page?

      @fetch.call(@next_cursor)
    end

    def each(&block)
      return enum_for(:each) unless block_given?

      page = self
      until page.nil?
        page.items.each(&block)
        page = page.next_page
      end
      # Ruby collection iterators return the receiver in block form.
      self
    end

    alias auto_paging_each each
  end
end
